azure-mgmt-resources/src/main/java/com/azure/management/resources/fluentcore/utils/ResourceManagerThrottlingPolicy.java

```diff
@@ -0,0 +1,132 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for
+ * license information.
+ */
+
+package com.azure.management.resources.fluentcore.utils;
+
+import com.azure.core.http.HttpPipelineCallContext;
+import com.azure.core.http.HttpPipelineNextPolicy;
+import com.azure.core.http.HttpRequest;
+import com.azure.core.http.HttpResponse;
+import com.azure.core.http.policy.HttpPipelinePolicy;
+import com.azure.core.util.DateTimeRfc1123;
+import com.azure.core.util.FluxUtil;
+import com.azure.core.util.logging.ClientLogger;
+import com.azure.management.resources.fluentcore.arm.ResourceUtils;
+import reactor.core.publisher.Mono;
+
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.OffsetDateTime;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * A Http Pipeline Policy for automatic retry when Azure Resource Manager is throttling because of too many read/write requests.
+ * <p>
+ * For each subscription and tenant, Azure Resource Manager limits read requests to 15,000 per hour and
+ *   write requests to 1,200 per hour. These limits apply to each Azure Resource Manager instance.
+ */
+public class ResourceManagerThrottlingPolicy implements HttpPipelinePolicy {
+    private static final ConcurrentMap<String, ReentrantLock> REENTRANT_LOCK_MAP = new ConcurrentHashMap<>();
+
+    @Override
+    public Mono<HttpResponse> process(HttpPipelineCallContext context, HttpPipelineNextPolicy next) {
+        final HttpRequest request = context.getHttpRequest();
+        String extractSubscriptionId = ResourceUtils.extractFromResourceId(request.getUrl().getPath(), "subscriptions");
+        if (extractSubscriptionId == null) {
+            extractSubscriptionId = "global";
+        }
+        final String subscriptionId = extractSubscriptionId;
+        REENTRANT_LOCK_MAP.putIfAbsent(subscriptionId, new ReentrantLock());
+        try {
+            synchronized (REENTRANT_LOCK_MAP.get(subscriptionId)) {
+                if (REENTRANT_LOCK_MAP.get(subscriptionId).isLocked()) {
+                    REENTRANT_LOCK_MAP.get(subscriptionId).wait();
```

Comment:
Please double check the code. I think v1 code is incorrect. Commonly no one calls `wait` on an ReentrantLock.

Reply:
`wait` makes thread waiting until `notifyAll`, it is not the method for ReentrantLock. Maybe change  it into `Object.wait()`?

Reply:
If one is using high level sync like ReentrantLock, one should never call low level sync method in Object (as well as synchronized block).

Please review what this class actually want to do. Normally one of lock or tryLock and then unlock should be enough to do the job.

Reply:
Since it is just wait the lock to be released (after sleep for retry), but don't need to lock again, do you mean it should be better to use `lock` then `unlock` immediately?

Reply:
If there is no better way you can do the `wait` on the object.

What I mean is that you need to think what is the proper way to implement the feature. And what it the proper way to test to make sure it is done correctly.

Having lock inside async method is already weird enough (and the unlock is in finally block of the method, does it happen before or after the request is complete?).
1st, is there another way to throttle it in async?
2nd, if you had to do the lock, is there a better way?

@yaohaizh

Reply:
I'd removed the lock, and deal with every 429 separately, since every subscription has several limitation (read, write, delete and so on)